### PR TITLE
Separate collector from inventory and targeted refresh

### PR DESCRIPTION
**Pull request by Ladas**
Separate collector from inventory, Collector class focus directly on collecting related data from a provider. Introduce targeted refresh.

### CODE REVIEW

File: app/models/manageiq/providers/amazon/cloud_manager/refresher.rb [hunk 1]
```diff
@@ -35,6 +44,29 @@ def parse_targeted_inventory(ems, _target, inventory)
     hashes
   end
 
+  def preprocess_targets
```
Comment (durandom) [inline app/models/manageiq/providers/amazon/cloud_manager/refresher.rb]: For now its ok, but I dont think we should allow this to be overriden

File: app/models/manageiq/providers/amazon/cloud_manager/refresher.rb [hunk 2]
```diff
@@ -35,6 +44,29 @@ def parse_targeted_inventory(ems, _target, inventory)
...
+    @targets_by_ems_id.each do |ems_id, targets|
+      if targets.any? { |t| t.kind_of?(ExtManagementSystem) }
+        ems             = @ems_by_ems_id[ems_id]
+        targets_for_log = targets.map { |t| "#{t.class} [#{t.name}] id [#{t.id}] " }
+        _log.info "Defaulting to full refresh for EMS: [#{ems.name}], id: [#{ems.id}], from targets: #{targets_for_log}" if targets.length > 1
+      end
+
+      # We want all targets of class EmsEvent to be merged into one target, so they can be refreshed together, otherwise
```
Comment (durandom) [inline app/models/manageiq/providers/amazon/cloud_manager/refresher.rb]: So, this thing about `EmsEvent` being a target. Is https://github.com/ManageIQ/manageiq/pull/13408 needed for this to work?

File: app/models/manageiq/providers/amazon/inventory/collectors.rb [hunk 3]
```diff
@@ -0,0 +1,126 @@
+class ManageIQ::Providers::Amazon::Inventory::Collectors
```
Comment (durandom) [inline app/models/manageiq/providers/amazon/inventory/collectors.rb]: I dont like this class being a plural. It should be `Collector`

File: app/models/manageiq/providers/amazon/inventory/collectors.rb [hunk 4]
```diff
@@ -0,0 +1,126 @@
...
+
+  def initialize(ems, target)
+    @ems     = ems
+    @target  = target
+    @options = Settings.ems_refresh[ems.class.ems_type]
+
+    initialize_inventory_sources
+  end
+
+  def initialize_inventory_sources
```
Comment (durandom) [inline app/models/manageiq/providers/amazon/inventory/collectors.rb]: So every collector has all this? Shouldnt it be per Subclass different?

File: app/models/manageiq/providers/amazon/inventory/targets.rb [hunk 5]
```diff
@@ -1,4 +1,9 @@
 class ManageIQ::Providers::Amazon::Inventory::Targets < ManageIQ::Providers::Amazon::Inventory
```
Comment (durandom) [inline app/models/manageiq/providers/amazon/inventory/targets.rb]: Same here, why is that a plural class name? And why does it inherit from `Inventory` ?

File: app/models/manageiq/providers/amazon/inventory/collectors.rb (hunk 3 above)
Comment (Ladas) [inline app/models/manageiq/providers/amazon/inventory/collectors.rb]: we can do singular, I have there Targets and Collectors

File: app/models/manageiq/providers/amazon/inventory/collectors.rb (hunk 4 above)
Comment (Ladas) [inline app/models/manageiq/providers/amazon/inventory/collectors.rb]: no, we need is always defined, so by default all data are []

File: app/models/manageiq/providers/amazon/inventory/targets.rb (hunk 5 above)
Comment (Ladas) [inline app/models/manageiq/providers/amazon/inventory/targets.rb]: The target is really an inventory here

We can revisit that Inventory could have target and collector

**Comment by durandom** [comment]
merging to move forward ontop of this

**Pull request merged**